src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -677,6 +678,95 @@ size_t runtime::gpu::CUDAEmitter::build_elementwise_n_to_1(const GPURuntimeConte
     return primitive_index;
 }
 
+size_t runtime::gpu::CUDAEmitter::build_reduce_window_helper(const GPURuntimeContext* ctx,
+                                                             const std::vector<std::string>& dtypes,
+                                                             const Shape& input_shape,
+                                                             const Shape& output_shape,
+                                                             const Shape& reduce_window_shape,
+                                                             const Strides& reduce_window_strides,
+                                                             const char* op,
+                                                             const char* kernel)
+{
+    // kernel_name is used to check if the cuda kernel has been previously compiled
+    size_t rank = input_shape.size();
+    std::stringstream kernel_name;
+    kernel_name << "reduce_window"
+                << "_" << op << "_" << join(dtypes, "_") << rank;
+
+    // hash is used to check if the emitted primitive already exists
+    std::stringstream ss;
+    ss << kernel_name.str() << "_s" << join(output_shape, "_");
+    auto hash = ss.str();
+
+    // if the primitive exists, we are done
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    // check if the kernel has already been compiled. if so, create
+    // a launch primitive for it based on the input tensor shape
+    // but do not recompile the kernel. otherwise, do it all:
+    // recompile the kernel and then create the primitive
+    auto compiled_kernel = ctx->compiled_kernel_pool->get(kernel_name.str());
+    if (compiled_kernel == nullptr)
+    {
+        codegen::CodeWriter writer;
+        CudaKernelBuilder::add_pod_typedefs(writer);
+        if (kernel)
+        {
+            CudaKernelBuilder::get_device_helper(writer, op, kernel, dtypes);
+        }
+
+        CudaKernelBuilder::get_reduce_window_op(writer, kernel_name.str(), op, dtypes, rank);
+
+        compiled_kernel = ctx->compiled_kernel_pool->set(kernel_name.str(), writer.get_code());
+    }
+    size_t nthreads = shape_size(output_shape);
+    auto input_strides = row_major_strides(input_shape);
+
+    std::shared_ptr<void*> input_strides_ptr = std::make_shared<void*>(
```

Comment:
Using shared_ptr here to avoid the need to free this allocation later? That won't work on a `void*` though. I would suggest using the new memory manager and the GPUAllocator as below. Replace lines 729 - 736 with:
```c++
    GPUAllocator allocator =
        external_function->get_primitive_emitter()->get_memory_allocator();
    size_t idx_input_strides = allocator.reserve_argspace(
        input_strides.data(), input_strides.size() * sizeof(size_t));
    size_t idx_output_shape = allocator.reserve_argspace(
        output_shape.data(), output_shape.size() * sizeof(size_t));
    size_t idx_reduce_window_shape = allocator.reserve_argspace(
        reduce_window_shape.data(), reduce_window_shape.size() * sizeof(size_t));
    size_t idx_reduce_window_strides = allocator.reserve_argspace(
        reduce_window_strides.data(), reduce_window_strides.size() * sizeof(size_t));
```

and then in the primitive get the device addresses with with:
```c++
    void* input_strides_d = runtime::gpu::invoke_memory_primitive(ctx, idx_input_strides);
    void* output_shape_d = runtime::gpu::invoke_memory_primitive(ctx, idx_output_shape);
    void* reduce_window_shape_d = runtime::gpu::invoke_memory_primitive(ctx, idx_reduce_window_shape);
    void* reduce_window_strides_d = runtime::gpu::invoke_memory_primitive(ctx, idx_reduce_window_strides);
```

Reply:
Here is actually by design, not a good design. 
It should have same lifetime as lambda function. It should not be deleted.

Reply:
memory allocation changes looks good.